### PR TITLE
Add text-to-speech sample (Matcha-TTS, FFT-free, Compiled Model GPU/CPU)

### PR DESCRIPTION
**Pull request by john-rocky**
## Matcha-TTS — on-device TTS with the Compiled Model API (FFT-free, GPU)

A new `compiled_model_api/text_to_speech/` sample running [Matcha-TTS](https://github.com/shivammehta25/Matcha-TTS) end-to-end on device: type text → synthesize → play. Complementary to the Kokoro TTS sample (#159, Interpreter API) — Matcha is the **FFT-free** path, so its heavy graphs ride the **CompiledModel GPU** delegate (a CFM acoustic model + HiFi-GAN time-domain vocoder, no FFT/iSTFT anywhere; spectral
vocoders like Kokoro need an FFT kernel the GPU delegate doesn't provide).

### Layout (matches the object-detection sample)
- `kotlin_cpu_gpu/android/` — the app (`MatchaSynthesizer.kt` host orchestration + 3 graphs, `MatchaG2P.kt`, `MainActivity.kt`).
- `conversion/` — the litert-torch conversion scripts (per the convert-code convention) + the on-device bug-isolation probes.
- `README.md`.

### Graphs (all `CompiledModel.create(...)`, fp16)
| Graph | Delegate |
|---|---|
| Text encoder | GPU |
| CFM decoder (×N ODE steps) | CPU¹ |
| HiFi-GAN vocoder | GPU |
| English G2P (DeepPhonemizer) | CPU |

Converted with litert-torch — per-graph tflite-vs-torch corr **1.000000**, end-to-end waveform corr ≥0.99 (GroupNorm→4D, Mish→SELECT-free softplus, ConvTranspose1d→ZeroStuffConvT1d, diffusers Attention→manual additive-masked with a runtime float mask = pad-to-max, SinusoidalPosEmb host-side).

¹ **The CFM decoder runs on the CompiledModel CPU delegate.** It converts GPU-clean and is exact on CPU, but on the Pixel 8a the Mali ML Drift GPU delegate **mis-fuses the decoder's transformer blocks at large activation magnitude** — the residual collapses (corr 0.006 vs CPU) when the block is fused in the full decoder graph, while the *same block as a standalone GPU graph is correct (corr 0.984)*: a graph-fusion bug, not an unsupported op (every op is verified correct on the GPU). The GPU vocoder dominates wall time, so the pipeline stays realtime (**RTF ~0.8 on Pixel 8a**). `conversion/probe_tx_standalone.py` is the minimal repro — flagging it here for the runtime team.

### G2P (espeak-free)
275k-entry espeak-IPA dictionary (OpenPhonemizer, Clear BSD) primary + DeepPhonemizer (MIT) on LiteRT CPU for OOV words → keithito 178-symbol set.

Models: [`mlboydaisuke/Matcha-TTS-LiteRT`](https://huggingface.co/mlboydaisuke/Matcha-TTS-LiteRT) (pushed to the device via `install_to_device.sh`; not committed).

**Draft** — pending a final Gradle build-verify pass on top of the device-verified app logic.

### CODE REVIEW

**Comment by smilingday** [comment]
I clean the markdown, to merge broken lines. Could we make sure the markdown read as normal?

@protobird-git could help take a quick look.

**Comment by john-rocky** [comment]
Thanks @smilingday! I've unwrapped the hard-wrapped paragraphs in both READMEs (`text_to_speech/README.md` and `conversion/README.md`) so each paragraph is a single line and renders as normal prose. Tables and code blocks are untouched, and the wording is unchanged — only the line breaks were merged. Pushed in 1e406ff.

File: compiled_model_api/text_to_speech/kotlin_cpu_gpu/android/app/src/main/java/com/google/ai/edge/examples/text_to_speech/MatchaG2P.kt [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+        }
+    }
+
+    /** Full text -> matcha symbol ids (blanks are interspersed later by the synthesizer).
+     *  Host-side text normalization: ALL-CAPS acronyms are spelled letter-by-letter (GPU ->
+     *  "gee pee you") and numbers are read as words (4090 -> "four thousand ninety"). */
+    fun phonemize(text: String): IntArray {
+        val ipa = StringBuilder()
+        var first = true
+        fun add(p: String) { if (p.isNotEmpty()) { if (!first) ipa.append(' '); ipa.append(p); first = false } }
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/kotlin_cpu_gpu/android/app/src/main/java/com/google/ai/edge/examples/text_to_speech/MatchaG2P.kt]: Could you format those code better, using Google style guide?
https://developer.android.com/kotlin/style-guide 

For here, break down those lines into mutliple lines will be better.

You can find more Google code styles here:
https://google.github.io/styleguide/

File: compiled_model_api/text_to_speech/conversion/convert_final.py
Comment (smilingday) [inline compiled_model_api/text_to_speech/conversion/convert_final.py:1]: Try to follow the style guide more:
https://google.github.io/styleguide/pyguide.html

In general, break lines to make those more readable for human.

More style guide:
https://google.github.io/styleguide/

File: compiled_model_api/text_to_speech/conversion/_stub.py [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# Copyright 2025 The Google AI Edge Authors. All Rights Reserved.
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/conversion/_stub.py]: 2006

File: compiled_model_api/text_to_speech/conversion/convert_final.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+  matcha_decoder_fp16.tflite   in: x,mu(1,80,Tm), t_emb(1,1024), ymask(1,1,Tm) -> v(1,80,Tm)
+  matcha_vocoder_fp16.tflite   in: mel(1,80,Tm) -> wav(1,1,Tm*256)
+  emb.npy            (178,192) phoneme embedding table (host lookup)
+  time_embed.npz     weights to compute t_emb on host (SinusoidalPosEmb + MLP)
+  config.json        symbols, MAX_TEXT/MAX_MEL, mel_mean/std, hop, length_scale, sr
+
+Verifies fp16 graphs end-to-end (host orchestration) vs torch reference on several
+sentences -> waveform corr.
+
+Run: ~/clipconv/bin/python convert_final.py [MAX_MEL]
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/conversion/convert_final.py]: what's "clipconv"?

File: compiled_model_api/text_to_speech/conversion/_stub.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Shared scipy/_propack + getsourcefile stub for macOS probes.
+
+Import this FIRST (before transformers / litert_torch) in any probe script:
+
+    import _stub  # noqa: F401
+
+Mirrors the guards baked into probe_convert.py so raw-module probes can run too.
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/conversion/_stub.py]: what's probe_convert.py?

File: compiled_model_api/text_to_speech/conversion/convert_final.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    for name, module, inputs in specs:
+        p32 = B.convert(module, inputs, os.path.join(ART, f"matcha_{name}.tflite"))
+        p16 = B.to_fp16(p32, os.path.join(ART, f"matcha_{name}_fp16.tflite"))
+        _, clean = B.opcheck(p16, name + "_fp16")
+        size32, size16 = os.path.getsize(p32) / 1e6, os.path.getsize(p16) / 1e6
+        print(f"  {name}: fp32 {size32:.1f}MB -> fp16 {size16:.1f}MB  GPU-clean={clean}")
+        fp16[name] = p16
+
+    from ai_edge_litert.interpreter import Interpreter
+    it_te = Interpreter(model_path=fp16["textenc"])
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/conversion/convert_final.py]: could we use CompiledModel python API?

File: compiled_model_api/text_to_speech/conversion/e2e_masked.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+import sys
+
+import numpy as np
+import torch
+import torch.nn as nn
+
+import build_matcha as B
+from e2e_matcha import LENGTH_SCALE, g2p_ids, generate_path, sequence_mask
+
+os.environ.setdefault("PHONEMIZER_ESPEAK_LIBRARY", "/opt/homebrew/lib/libespeak-ng.dylib")
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/conversion/e2e_masked.py]: will those scripts only work on MacOS?
Could those easy to be able to run Linux too?

File: compiled_model_api/text_to_speech/kotlin_cpu_gpu/android/app/src/main/java/com/google/ai/edge/examples/text_to_speech/MatchaSynthesizer.kt [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+            cumulative[i] = total
+        }
+        val melLength = minOf(maxOf(total.toInt(), 1), MAX_MEL)
+
+        val muY = FloatArray(N_FEATS * MAX_MEL)
+        var phonemeIndex = 0
+        for (frame in 0 until melLength) {
+            while (phonemeIndex < MAX_TEXT - 1 && cumulative[phonemeIndex] <= frame) phonemeIndex++
+            for (c in 0 until N_FEATS) muY[c * MAX_MEL + frame] = mu[c * MAX_TEXT + phonemeIndex]
+        }
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/kotlin_cpu_gpu/android/app/src/main/java/com/google/ai/edge/examples/text_to_speech/MatchaSynthesizer.kt:138]: In general, the code styles suggest to break lines into more clear blocks, and you can write things like below, which will be easy to follow.

ditto for many other things. Don't mix a lot of codes together, which will be hard to follow.

===

for (frame in 0 until melLength) {
    while (phonemeIndex < MAX_TEXT - 1 && cumulative[phonemeIndex] <= frame) {
        phonemeIndex++
    }
    for (c in 0 until N_FEATS) {
        muY[c * MAX_MEL + frame] = mu[c * MAX_TEXT + phonemeIndex]
    }
}

File: compiled_model_api/text_to_speech/conversion/_stub.py (hunk 2 above)
Comment (john-rocky) [reply compiled_model_api/text_to_speech/conversion/_stub.py]: Good catch — bumped the copyright header to 2026 across the new files in this PR (4f3458d).

**Comment by john-rocky** [comment]
Thanks for the detailed review @smilingday — all points addressed:

- **Readability / explicit blocks** (`MatchaG2P.kt`, `MatchaSynthesizer.kt`): expanded the single-statement `if`/`for`/`while` bodies onto their own braced lines so each step reads clearly, per your length-regulator example (9bb7ddc).
- **CompiledModel Python API**: the parity runs now go through the LiteRT `CompiledModel` API (`from_file` + `run_by_index`) and the GPU-compat op scan reads the `.tflite` flatbuffer directly instead of instantiating an `Interpreter` (422d405; finished the migration in `e2e_masked.py` in 04f8edd).
- **Cross-platform (Linux too)**: espeak is now resolved via `ctypes.util.find_library` plus common macOS/Linux locations instead of the hardcoded Homebrew `.dylib`, and the env var is left unset if none is found (04f8edd).
- **`clipconv`**: that was my local venv — replaced `~/clipconv/bin/python` with `python` in the usage docstrings, and removed the dead `probe_convert.py` reference in `_stub.py` (04f8edd).
- **Copyright year**: set to 2026 for the new files (4f3458d).

Re-verified end to end: the Python pipeline synthesizes correctly through the CompiledModel API on the fp16 graphs (`kotlin_replica.py` → 51/53 phonemes, ~292 frames each), and the Android app still runs on a Pixel 8a (RTF ~0.75, all-GPU).

File: compiled_model_api/text_to_speech/conversion/build_matcha.py [hunk 8]
```diff
@@ -0,0 +1,648 @@
...
+# limitations under the License.
+
+"""Matcha-TTS -> CompiledModel GPU: real-weight conversion + parity checks.
+
+Loads matcha_ljspeech.ckpt + generator_v1 (HiFi-GAN T2 v1), applies the proven
+re-authoring recipe to the 3 heavy graphs (text encoder / CFM decoder / HiFi-GAN
+vocoder), converts each with litert-torch, and verifies:
+  (a) re-authored torch == original torch  (mask-drop is a no-op at full length)
+  (b) tflite == re-authored torch          (conversion fidelity)
+  (c) end-to-end waveform: full tflite-orchestrated pipeline vs torch synthesise()
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/conversion/build_matcha.py]: This line exceeds 80 characters. Please wrap it or break it up to follow the Google Python Style Guide.

Comment (smilingday) [reply compiled_model_api/text_to_speech/conversion/build_matcha.py]: https://google.github.io/styleguide/pyguide.html#32-line-length

File: compiled_model_api/text_to_speech/conversion/build_matcha.py [hunk 9]
```diff
@@ -0,0 +1,648 @@
...
+    with torch.no_grad():
+        run_dummy()
+    for hook in hooks:
+        hook.remove()
+    return lengths
+
+
+def reauth_decoder(dec, T):
+    """Host-side time embed; GN->4D, Mish->softplus, ConvT->ZeroStuff, drop attn mask.
+    Graph input = (x(1,80,T), mu(1,80,T), t_emb(1,1024)); output v(1,80,T)."""
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/conversion/build_matcha.py]: https://google.github.io/styleguide/pyguide.html#doc-function-args

File: compiled_model_api/text_to_speech/kotlin_cpu_gpu/android/app/src/main/java/com/google/ai/edge/examples/text_to_speech/MatchaSynthesizer.kt [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+
+        // Text encoder (GPU) -> mu[1,80,T], logw[1,1,T].
+        textEncoderInputs[0].writeFloat(embedded)
+        textEncoderInputs[1].writeFloat(textMask)
+        textEncoder.run(textEncoderInputs, textEncoderOutputs)
+        val mu = textEncoderOutputs[0].readFloat() // [N_FEATS * MAX_TEXT], channel-major
+        val logDurations = textEncoderOutputs[1].readFloat() // [MAX_TEXT]
+
+        // Durations -> cumulative -> length regulator (mu_y).
+        val durations = FloatArray(MAX_TEXT) { ceil(exp(logDurations[it]) * textMask[it]) * LENGTH_SCALE }
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/kotlin_cpu_gpu/android/app/src/main/java/com/google/ai/edge/examples/text_to_speech/MatchaSynthesizer.kt]: this line is > 100 chars.

In general, you can break down codes into more readable styles like

val durations = FloatArray(MAX_TEXT) {
    ceil(exp(logDurations[it]) * textMask[it]) * LENGTH_SCALE
}

File: compiled_model_api/text_to_speech/conversion/probe_decoder_taps.py [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+import torch
+import torch.nn as nn
+from einops import pack, rearrange
+
+import build_matcha as B
+
+T = 512
+
+
+def dbg_forward(self, x, mask, mu, t):
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/conversion/probe_decoder_taps.py:39]: Could define Args:

https://google.github.io/styleguide/pyguide.html#doc-function-args

File: compiled_model_api/text_to_speech/conversion/probe_decoder_taps.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+        b, c, length = m.shape
+        return m.reshape(b, c, length // 2, 2)[:, :, :, 0]
+
+    taps = {}
+    t = self.time_embeddings(t)
+    t = self.time_mlp(t)
+    x = pack([x, mu], "b * t")[0]
+    hiddens = []
+    masks = [mask]
+    for block_index, (resnet, transformer_blocks, downsample) in enumerate(self.down_blocks):
```
Comment (smilingday) [inline compiled_model_api/text_to_speech/conversion/probe_decoder_taps.py]: Add empty lines between complex code blocks will make it easy to follow.

**Review by smilingday — COMMENTED** [review]
Python line suggestion is 80 chars, while Kotlin is 100.

Could add Args to explain the argments.

Try to add more empty lines, or break long lines to multiple lines, to let easy to follow.

But in general, let's go ahead to submit this, and OK for now. We can polish in future.

**Comment by john-rocky** [comment]
Thanks @smilingday — both points addressed:

- **80-column limit**: wrapped every line in `conversion/*.py` to 80 columns per the Google Python Style Guide (130 long lines across the 9 scripts, including the flagged `build_matcha.py:22` docstring). Code lines are split at meaningful points and docstrings reflowed — no string contents or behavior changed.
- **Function docstrings ([doc-function-args](https://google.github.io/styleguide/pyguide.html#doc-function-args))**: added Google-format `Args:` / `Returns:` sections to the module-level function docstrings in the conversion scripts, including the flagged `reauth_decoder`.

Pushed in 01f2943; `python -m py_compile` is green on all 9 files and the app still builds. I'm also applying the same 80-column + docstring pass uniformly to the conversion scripts in all my other open PRs, so this won't need to be re-flagged per PR.

**Review by smilingday — DISMISSED** [review]

**Comment by john-rocky** [comment]
Thank you for the approval @smilingday! I've also addressed the remaining polish points in 90320c3 rather than leaving them for later:

- **Kotlin 100-column limit**: broke the six lines that exceeded 100 columns in `MatchaSynthesizer.kt` / `MatchaG2P.kt`, using your suggested `FloatArray(MAX_TEXT) { ... }` builder style for the duration computation. App still builds.
- **Args on `probe_decoder_taps.py`**: the `dbg_forward` docstring got a full `Args:`/`Returns:` section in the earlier 80-column commit (01f2943), together with all other module-level functions.
- **Empty lines between complex blocks**: added blank lines with short section comments (`# Down blocks.` / `# Mid blocks.` / `# Up blocks.` / `# Final projection.`) to the dense decoder-forward copies in `probe_decoder_taps.py` and `build_matcha.py`.

On the **Build-and-Test-Linux failure**: the failing step is the bazel build of `//compiled_model_api/semantic_similarity/build_from_source:semantic_similarity`, which errors while resolving the external repo (`cannot load '@litert_archive//litert:tensorflow_source_rules.bzl': no such file`). This PR adds no bazel or C++ targets, and the same failure occurred on the previous commit as well, so it looks like a repository-wide breakage of that workflow's external dependency rather than something introduced here.

**Review by smilingday — DISMISSED** [review]

**Comment by smilingday** [comment]
this is fixed, and could you update the PR to catch the latest change?

**Comment by aravindmurali22** [comment]
Hi @john-rocky , could you please pull the latest changes from main branch into your PR? the latest PR will fix the broken github workflows. Thank you.

**Review by smilingday — APPROVED** [review]

**Review by smilingday — APPROVED** [review]

**Comment by smilingday** [comment]
Please reply and resolve my comment, when you have address it.
